XR_APILAYER_NOVENDOR_toolkit/fsr.cpp

```diff
@@ -0,0 +1,216 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "pch.h"
+
+#include "shader_utilities.h"
+#include "factories.h"
+#include "interfaces.h"
+#include "log.h"
+
+#define A_CPU
+#include <ffx_a.h>
+#include <ffx_fsr1.h>
+
+namespace toolkit {
+
+    extern std::string dllHome;
+
+} // namespace toolkit
+
+namespace {
+
+    using namespace toolkit;
+    using namespace toolkit::config;
+    using namespace toolkit::graphics;
+    using namespace toolkit::log;
+
+    
+    struct FSRConstants {
+        uint32_t Const0[4];
+        uint32_t Const1[4];
+        uint32_t Const2[4];
+        uint32_t Const3[4];
+        uint32_t Const4[4];
+    };
+
+    class FSRUpscaler : public IUpscaler {
+      public:
+        FSRUpscaler(std::shared_ptr<IConfigManager> configManager,
+                    std::shared_ptr<IDevice> graphicsDevice,
+                    uint32_t outputWidth,
+                    uint32_t outputHeight)
+            : m_configManager(configManager)
+            , m_device(graphicsDevice), m_outputWidth(outputWidth), m_outputHeight(outputHeight) {
+
+            // The upscaling factor is only read upon initialization of the session. It cannot be changed after.
+            std::tie(m_inputWidth, m_inputHeight) =
+                utilities::GetScaledDimensions(m_outputWidth, m_outputHeight, configManager->getValue(SettingScaling), 2);
+
+            if (m_inputWidth != m_outputWidth || m_inputHeight != m_outputHeight) {
+                initializeScaler();
+            } else {
+                initializeSharpen();
+            }
+
+            update();
+        }
+
+        void update() override {
+            if (m_configManager->hasChanged(SettingSharpness)) {
+                const auto sharpness = m_configManager->getValue(SettingSharpness) / 100.f;
+                const auto attenuation = 1.f - AClampF1(sharpness, 0, 1);
+
+                FSRConstants config = {};
+                if (!m_isSharpenOnly) {
+                    FsrEasuCon(config.Const0, config.Const1, config.Const2, config.Const3,
+                               static_cast<AF1>(m_inputWidth ), static_cast<AF1>(m_inputHeight ),
+                               static_cast<AF1>(m_inputWidth ), static_cast<AF1>(m_inputHeight ),
+                               static_cast<AF1>(m_outputWidth), static_cast<AF1>(m_outputHeight));
+                }
+
+                FsrRcasCon(config.Const4, static_cast<AF1>(attenuation));
+
+                // implementation is using a shader compilation define for this
+                //config.Const4[3] = (hdr != NISHDRMode::None) ? 1 : 0;
```

Comment:
Is this comment stale? It's referring to NIS.

Reply:
The AMD FSR sample is using a value in the constant buffer to correct the output color accordingly. I've replaced the constant with a shader compilation time define for now because the project code is not HDR aware yet, but in case we do in the future and we need to restore runtime shader handling of this case, I've left the comment as a "reminder". Would making it a TODO entry ok?

Reply:
Yes let's make it a TODO. I was a bit confused by the use of the NIS-specific enum.